Missing parenthesis in help message

### Code

```Rust
struct Foo;

struct Bar;

impl From<&Foo> for Bar {
    fn from(foo: &Foo) -> Bar {
        Bar
    }
}

fn main() {
    let foo = Foo;
    let b: Bar = foo.into();
}
```


### Current output

```Shell
error[E0277]: the trait bound `Foo: Into<_>` is not satisfied
  --> src/main.rs:13:22
   |
13 |     let b: Bar = foo.into();
   |                      ^^^^ the trait `~const Into<_>` is not implemented for `Foo`
   |
   = note: required for `Foo` to implement `Into<Bar>`
help: consider borrowing here
   |
13 |     let b: Bar = &foo.into();
   |                  +

For more information about this error, try `rustc --explain E0277`.
error: could not compile `reference-loop` due to previous error
```


### Desired output

```Shell
error[E0277]: the trait bound `Foo: Into<_>` is not satisfied
  --> src/main.rs:13:22
   |
13 |     let b: Bar = foo.into();
   |                      ^^^^ the trait `~const Into<_>` is not implemented for `Foo`
   |
   = note: required for `Foo` to implement `Into<Bar>`
help: consider borrowing here
   |
13 |     let b: Bar = (&foo).into();
   |                   +

For more information about this error, try `rustc --explain E0277`.
error: could not compile `reference-loop` due to previous error
```


### Rationale and extra context

_No response_

### Other cases

_No response_

### Anything else?

The help message suggests borrowing `foo.into()`, which obviously doesn't fix the issue (it leads to a series of other misleading errors). This happens using `rustc 1.68.0 (2c8cc3432 2023-03-06)`.

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"MU001999"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs
+++ b/compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs
@@ -1358,6 +1358,34 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
                             Applicability::MaybeIncorrect,
                         );
                     } else {
+                        let is_mut = mut_ref_self_ty_satisfies_pred || ref_inner_ty_mut;
+                        let sugg_prefix = format!("&{}", if is_mut { "mut " } else { "" });
+                        let sugg_msg = &format!(
+                            "consider{} borrowing here",
+                            if is_mut { " mutably" } else { "" }
+                        );
+
+                        // Issue #109436, we need to add parentheses properly for method calls
+                        // for example, `foo.into()` should be `(&foo).into()`
+                        if let Ok(snippet) = self
+                            .tcx
+                            .sess
+                            .source_map()
+                            .span_to_snippet(self.tcx.sess.source_map().next_point(span))
+                        {
+                            if snippet == "." {
+                                err.multipart_suggestion_verbose(
+                                    sugg_msg,
+                                    vec![
+                                        (span.shrink_to_lo(), format!("({}", sugg_prefix)),
+                                        (span.shrink_to_hi(), ")".to_string()),
+                                    ],
+                                    Applicability::MaybeIncorrect,
+                                );
+                                return true;
+                            }
+                        }
+
                         // Issue #104961, we need to add parentheses properly for compond expressions
                         // for example, `x.starts_with("hi".to_string() + "you")`
                         // should be `x.starts_with(&("hi".to_string() + "you"))`
@@ -1374,14 +1402,7 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
                             _ => false,
                         };
 
-                        let is_mut = mut_ref_self_ty_satisfies_pred || ref_inner_ty_mut;
                         let span = if needs_parens { span } else { span.shrink_to_lo() };
-                        let sugg_prefix = format!("&{}", if is_mut { "mut " } else { "" });
-                        let sugg_msg = &format!(
-                            "consider{} borrowing here",
-                            if is_mut { " mutably" } else { "" }
-                        );
-
                         let suggestions = if !needs_parens {
                             vec![(span.shrink_to_lo(), format!("{}", sugg_prefix))]
                         } else {

--- a/compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs
+++ b/compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs
@@ -1367,12 +1367,9 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
 
                         // Issue #109436, we need to add parentheses properly for method calls
                         // for example, `foo.into()` should be `(&foo).into()`
-                        if let Ok(snippet) = self
-                            .tcx
-                            .sess
-                            .source_map()
-                            .span_to_snippet(self.tcx.sess.source_map().next_point(span))
-                        {
+                        if let Ok(snippet) = self.tcx.sess.source_map().span_to_snippet(
+                            self.tcx.sess.source_map().span_look_ahead(span, Some("."), Some(50)),
+                        ) {
                             if snippet == "." {
                                 err.multipart_suggestion_verbose(
                                     sugg_msg,

--- a/tests/ui/suggestions/issue-109436.rs
+++ b/tests/ui/suggestions/issue-109436.rs
@@ -1,0 +1,13 @@
+struct Foo;
+struct Bar;
+
+impl From<&Foo> for Bar {
+    fn from(foo: &Foo) -> Bar {
+        Bar
+    }
+}
+
+fn main() {
+    let foo = Foo;
+    let b: Bar = foo.into(); //~ ERROR E0277
+}

--- a/tests/ui/suggestions/issue-109436.stderr
+++ b/tests/ui/suggestions/issue-109436.stderr
@@ -1,0 +1,15 @@
+error[E0277]: the trait bound `Foo: Into<_>` is not satisfied
+  --> $DIR/issue-109436.rs:12:22
+   |
+LL |     let b: Bar = foo.into();
+   |                      ^^^^ the trait `~const Into<_>` is not implemented for `Foo`
+   |
+   = note: required for `Foo` to implement `Into<Bar>`
+help: consider borrowing here
+   |
+LL |     let b: Bar = (&foo).into();
+   |                  ++   +
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0277`.

